substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java

```diff
@@ -0,0 +1,548 @@
+import org.graalvm.compiler.nodes.java.LoadFieldNode;
+import org.graalvm.compiler.nodes.java.MethodCallTargetNode;
+import org.graalvm.compiler.nodes.java.NewArrayNode;
+import org.graalvm.compiler.nodes.java.NewInstanceNode;
+import org.graalvm.compiler.nodes.java.StoreFieldNode;
+import org.graalvm.compiler.nodes.spi.UncheckedInterfaceProvider;
+import org.graalvm.compiler.nodes.type.StampTool;
+import org.graalvm.compiler.options.Option;
+import org.graalvm.compiler.options.OptionValues;
+import org.graalvm.compiler.phases.OptimisticOptimizations;
+import org.graalvm.compiler.phases.util.Providers;
+import org.graalvm.compiler.word.WordTypes;
+
+import com.oracle.graal.pointsto.BigBang;
+import com.oracle.graal.pointsto.meta.AnalysisMethod;
+import com.oracle.graal.pointsto.meta.AnalysisType;
+import com.oracle.graal.pointsto.meta.HostedProviders;
+import com.oracle.graal.pointsto.util.CompletionExecutor.DebugContextRunnable;
+import com.oracle.svm.core.annotate.NeverInline;
+import com.oracle.svm.core.annotate.NeverInlineTrivial;
+import com.oracle.svm.core.annotate.RestrictHeapAccess;
+import com.oracle.svm.core.annotate.Uninterruptible;
+import com.oracle.svm.core.option.HostedOptionKey;
+import com.oracle.svm.core.util.VMError;
+import com.oracle.svm.hosted.meta.HostedMethod;
+import com.oracle.svm.hosted.phases.AnalysisGraphBuilderPhase.AnalysisBytecodeParser;
+import com.oracle.svm.hosted.phases.NativeImageInlineDuringParsingPlugin.CallSite;
+import com.oracle.svm.hosted.phases.NativeImageInlineDuringParsingPlugin.InvocationResult;
+import com.oracle.svm.hosted.phases.NativeImageInlineDuringParsingPlugin.InvocationResultInline;
+import com.oracle.svm.hosted.phases.SharedGraphBuilderPhase.SharedBytecodeParser;
+
+import jdk.vm.ci.meta.ResolvedJavaMethod;
+
+/**
+ * Plugin for specifying what is inlined during graph parsing before the static analysis. This
+ * plugin inline methods which folds to constant and is also notified {@link #notifyBeforeInline}
+ * and {@link #notifyAfterInline} the inlining.
+ */
+public class NativeImageInlineDuringParsingPlugin implements InlineInvokePlugin {
+
+    public static class Options {
+        @Option(help = "Inline methods which folds to constant during parsing before the static analysis.") public static final HostedOptionKey<Boolean> InlineBeforeAnalysis = new HostedOptionKey<>(
+                false);
+
+    }
+
+    private final boolean analysis;
+    private final HostedProviders providers;
+
+    /* the map in which we place the result of a inlining decision */
+    private static final ConcurrentHashMap<AnalysisMethod, ConcurrentHashMap<Integer, InvocationResult>> dataInline = new ConcurrentHashMap<>();
+
+    public NativeImageInlineDuringParsingPlugin(boolean analysis, HostedProviders providers) {
+        this.analysis = analysis;
+        this.providers = providers;
+    }
+
+    /**
+     * Determines whether a call to a given method is to be inlined.
```

Comment:
This comment, on the other hand, is completely useless as all of it is clear from the signature.

Reply:
Removed.